### PR TITLE
Do not crash on intersecting virtual parts

### PR DESCRIPTION
**Pull request by tavplubix**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Bug Fix

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
 - If ZooKeeper connection was lost and replica was cloned after restoring the connection, its replication queue might contain outdated entries. It's fixed.
 - Fixed crash when replication queue contains intersecting virtual parts. It may rarely happen if some data part was lost. Print error in log instead of terminating.

### CODE REVIEW

File: src/Storages/MergeTree/MergeTreeData.cpp [hunk 1]
```diff
@@ -4897,7 +4897,7 @@ void MergeTreeData::removeQueryId(const String & query_id) const
 {
     std::lock_guard lock(query_id_set_mutex);
     if (query_id_set.find(query_id) == query_id_set.end())
-        LOG_WARNING(log, "We have query_id removed but it's not recorded. This is a bug");
+        throw Exception(ErrorCodes::LOGICAL_ERROR, "We have query_id removed but it's not recorded. This is a bug");
```
Comment (amosbird) [inline src/Storages/MergeTree/MergeTreeData.cpp:4900]: The method is used in dtor. Shall we still throw?

Comment (tavplubix) [reply src/Storages/MergeTree/MergeTreeData.cpp:4900]: It's ok to abort on logical error in some cases. But we can replace `throw ...` with `assert(false)` to abort in debug builds only.

Comment (amosbird) [reply src/Storages/MergeTree/MergeTreeData.cpp:4900]: Yeah, that would be consistent with other code.

**Comment by tavplubix** [comment]
I realized that solution with special `PART_IS_LOST` log entry does not work if some replica was cloned during execution of such entry:

1. Let replication queue of replica `r1` contains `MERGE_PARTS` entry to merge `all_1_1_0` and `all_2_2_0` into `all_1_2_1` and part `all_1_1_0` is lost forever, so merge cannot be executed and must be cancelled. 
2. Replica `r1` creates log entry `PART_IS_LOST` for part `all_1_2_1`. 
3. Another replica `r2` is cloning replica `r1` and [coping it's queue](https://github.com/ClickHouse/ClickHouse/blob/master/src/Storages/StorageReplicatedMergeTree.cpp#L2721-L2728). `r2` successfully gets `MERGE_PARTS`.
4. `r1` is executing `PART_IS_LOST`. `r1` removes both `MERGE_PARTS` (as cancelled) and `PART_IS_LOST` (as processed) entries from own queue.
5. `r2` is still copying `r1`s queue, but it cannot get `PART_IS_LOST` entry. Replication queue of `r2` will contain `MERGE_PARTS` (for `all_1_2_1`), but will not contain the corresponding `PART_IS_LOST`. `r2` does not know yet, that `all_1_1_0` is lost and merge is impossible to execute
6. `r1` creates new entry to merge `all_2_2_0` and `all_3_3_0` into `all_2_3_1`
7. `r2` pulls log to own queue, but it's queue already contain merge entry for `all_1_2_1`. Merge entry for `all_2_3_1` intersects it.

Seems like we can fix it by introducing something like counter of executed `PART_IS_LOST` entries for each replica and retry `cloneReplica(...)` if source replica has changed its counter, but current solution is already complicated and extra counter will make it even worse.

**Comment by amosbird** [comment]
> Replica r1 creates log entry PART_IS_LOST for part all_1_2_1.

How about we generating an empty `all_1_1_0` part when it's lost forever instead of canceling planned merges with `PART_IS_LOST`?

**Comment by tavplubix** [comment]
> > Replica r1 creates log entry PART_IS_LOST for part all_1_2_1.
> 
> How about we generating an empty `all_1_1_0` part when it's lost forever instead of canceling planned merges with `PART_IS_LOST`?

It's possibly a good idea, but it's not that simple too. Some considerations:
- To replace lost part with empty part we have to change part checksum and it breaks invariant "parts are immutable". However, it's possibly not a problem, because no replicas have such part (so we don't need to change anything inside `replica_path/parts` where checksums are stored) and it doesn't matter if lost part is immutable.
- It's definitely safe to replace a part with empty one only if the part has 0 level and mutation version (i.e. part was just inserted). I'm not sure about other cases
- Maybe we should make some changes in replication queue reordering (`shouldExecuteLogEntry(...)`, `prefer_fetch_merged_part` logic, etc) to allow execution of `GET_PART` entry when part is probably lost (and we are going to replace it with empty part), but it's covered by `MERGE_PARTS` or `MUTATE_PART` entry.
- It may fix some particular case of "some merges may stuck" issue (#10368).

**Comment by amosbird** [comment]
> It's definitely safe to replace a part with empty one only if the part has 0 level and mutation version (i.e. part was just inserted). I'm not sure about other cases

Interesting thinking. It might indeed introduce unexpected user experience when some lower level parts are still active and suddenly becomes invisible due to introducing a high level empty fake part.

Now we might need to proper define a standard behavior when a part is lost: Shall we allow `all_1_1_0_1` to serve as a backup part for `all_1_1_0_2`

File: src/Storages/MergeTree/ReplicatedMergeTreeQueue.h [hunk 2]
```diff
@@ -92,9 +92,6 @@ class ReplicatedMergeTreeQueue
     using FuturePartsSet = std::map<String, LogEntryPtr>;
     FuturePartsSet future_parts;
 
-    /// Index of the first log entry that we didn't see yet.
-    Int64 log_pointer = 0;
```
Comment (tavplubix) [inline src/Storages/MergeTree/ReplicatedMergeTreeQueue.h:96]: It was unused (value was never read)

File: src/Storages/MergeTree/ActiveDataPartSet.cpp [hunk 3]
```diff
@@ -56,15 +58,18 @@ bool ActiveDataPartSet::add(const String & name, Strings * out_replaced_parts)
     /// Let's go to the right.
     while (it != part_info_to_name.end() && part_info.contains(it->first))
     {
-        if (part_info == it->first)
-            throw Exception(ErrorCodes::LOGICAL_ERROR, "Unexpected duplicate part {}. It is a bug.", name);
+        assert(part_info != it->first);
```
Comment (alesapin) [inline src/Storages/MergeTree/ActiveDataPartSet.cpp:61]: Logging was lost?

**Review by alesapin — APPROVED** [review]

File: src/Storages/MergeTree/ActiveDataPartSet.cpp (hunk 3 above)
Comment (tavplubix) [reply src/Storages/MergeTree/ActiveDataPartSet.cpp:61]: It was redundant, see https://github.com/ClickHouse/ClickHouse/pull/24723#issuecomment-850649761

**Comment by tavplubix** [comment]
Functional stateless tests (address) - 01130_in_memory_parts - `Exception happened during execution of mutation 'mutation_5.txt' with part 'all_1_1_0_4' reason: 'Code: 236, e.displayText() = DB::Exception: Cancelled mutating parts` - seems like some test cancels merges and mutations globally. We need to find the test and fix it
Functional stateless tests (release, DatabaseReplicated) - known issue, already fixed in master
Functional stateless tests flaky check (address) - `Test runs too long` - ok to ignore, no new failures
Performance - looks unrelated
SQLancer test - SQLancer works incorrectly
Yandex third-party checks (only for Yandex employees) - something related to `Context`, was broken in master

**Pull request merged**